File: lollys-lotto-ts-sdk/src/codegen/instructions/claimUserRewards.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface ClaimUserRewardsArgs {
  amountToBeClaimed: BN
}

export interface ClaimUserRewardsAccounts {
  user: PublicKey
  userUsdcTokenAccount: PublicKey
  userMetadata: PublicKey
  /** Mint address of the USDC token */
  usdcMint: PublicKey
  userRewardsVault: PublicKey
  eventEmitter: PublicKey
  tokenProgram: PublicKey
}

export const layout = borsh.struct([borsh.u64("amountToBeClaimed")])

export function claimUserRewards(
  args: ClaimUserRewardsArgs,
  accounts: ClaimUserRewardsAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.user, isSigner: true, isWritable: false },
    {
      pubkey: accounts.userUsdcTokenAccount,
      isSigner: false,
      isWritable: true,
    },
    { pubkey: accounts.userMetadata, isSigner: false, isWritable: true },
    { pubkey: accounts.usdcMint, isSigner: false, isWritable: false },
    { pubkey: accounts.userRewardsVault, isSigner: false, isWritable: true },
    { pubkey: accounts.eventEmitter, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([219, 205, 212, 97, 101, 221, 21, 245])
  const buffer = Buffer.alloc(1000)
  const len = layout.encode(
    {
      amountToBeClaimed: args.amountToBeClaimed,
    },
    buffer
  )
  const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len)
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}

File: lollys-lotto-rust-sdk/src/instructions/user/claim_user_rewards.rs
use crate::instructions::*;

pub fn claim_user_rewards(
    amount_to_be_claimed: u64,
    user: &Pubkey,
    user_usdc_token_account: &Pubkey,
    user_metadata: &Pubkey,
    usdc_mint: &Pubkey,
    user_rewards_vault: &Pubkey,
    event_emitter: &Pubkey,
) -> Instruction {
    let data = lollys_lotto::instruction::ClaimUserRewards {
        amount_to_be_claimed,
    }
    .data();

    let accounts = lollys_lotto::accounts::ClaimUserRewards {
        user: *user,
        user_usdc_token_account: *user_usdc_token_account,
        user_metadata: *user_metadata,
        usdc_mint: *usdc_mint,
        user_rewards_vault: *user_rewards_vault,
        event_emitter: *event_emitter,
        token_program: token::ID,
    }
    .to_account_metas(None);

    Instruction {
        program_id: lollys_lotto::id(),
        accounts,
        data,
    }
}